### ISSUE TITLE
Runs twice then: Uncaught (in promise) Error: Invalid LatLng object: (NaN, NaN)

### ISSUE THREAD

**Issue by Terrycoco** [IvanSanchez/Leaflet.Polyline.SnakeAnim#29]
Snaking animation runs a few times then get following error stack:

```
leaflet-src.js:1370 Uncaught (in promise) Error: Invalid LatLng object: (NaN, NaN)
    at new LatLng (leaflet-src.js:1370)
    at Object.unproject (leaflet-src.js:1670)
    at Object.pointToLatLng (leaflet-src.js:1511)
    at NewClass.unproject (leaflet-src.js:3908)
    at NewClass.layerPointToLatLng (leaflet-src.js:3916)
    at NewClass.containerPointToLatLng (leaflet-src.js:3973)
    at NewClass._snakeForward (polyline.snakeanim.js:122)
    at NewClass._snake (polyline.snakeanim.js:76)
    at NewClass.snakeIn (polyline.snakeanim.js:60)
    at snake (SnakeMap.js:174)

```
I checked my latlngs and they are not passing Nan in the array.

**Comment by IvanSanchez** [comment]
Can you publish a fiddle/codepen/[plunkr](https://leafletjs.com/edit) that displays this behaviour?

**Comment by nathanh2011** [comment]
Hi, I get the same issue now all of a sudden. My array has complete and proper lat/lngs.

I have narrowed that the error is occurring in `_snakeForward` when calling `this._map.latLngToContainerPoint()`. I think it is caused by points that are the same being passed in for example:

```
[-35.3123241, 149.1461274],
[-35.3123241, 149.1461274],
[-35.3123241, 149.1461274],
[-35.3123241, 149.1461274],
[-35.3123241, 149.1461274],
[-35.3123241, 149.1461274],
[-35.3123241, 149.1461274],
[-35.3153949, 149.1503217],
[-35.3145958, 149.1504801],
[-35.3146478, 149.1504697]
```
Which I thought I could fix by eliminating the duplicated before adding them into your plugin but it looks like in _snakeForward when calling the points it loses some accuracy and goes to 5 decimal places and I am finding points that are close by are causing the same issue. For example this block also fails with the error:

```
[-35.2613033, 149.1343751],
[-35.2612639, 149.1343781],
[-35.261261, 149.13443],
[-35.2612579, 149.1343983],
[-35.26127, 149.1343797],
[-35.2613504, 149.1343453],
```

Although these are all distinct points, inside _snakeForward they are seen as (currPoint - nextPoint):
```
LatLng(-35.26103, 149.13432) - LatLng(-35.2613, 149.13438)
LatLng(-35.2613, 149.13438) - LatLng(-35.26126, 149.13438)
LatLng(-35.26126, 149.13438) - LatLng(-35.26126, 149.13443)
LatLng(-35.26126, 149.13443) - LatLng(-35.26126, 149.1344)
Error: Invalid LatLng object: (NaN, NaN)
```

So I am not sure if it is if a lat or lon match or if both need to be the same to cause the issue, but it looks like a rounding happening somewhere is causing non-distinct points to be seen as the same which looks to be causing the problem.

**Comment by nathanh2011** [comment]
With further testing I have found the map zoom level determines how sensitive this issue is. If I am zoomed out then the issue is much more immediate as points as drawn close together. As a quick fix I have found that if I set my zoom level to 16 while it is snaking, all my points render fine and at the end I reset the zoom back out to 14.

**Comment by Terrycoco** [comment]
I found a work around by accounting for the "forward" variable in _snakeForward to not be 0;

 ```
 _snakeForward: function(forward) {
forward = (forward == 0 ? 0.00000000001 : forward);
...
```

Perhaps you can find a more elegant solution

**Comment by nathanh2011** [comment]
>  I found a work around by accounting for the "forward" variable in _snakeForward to not be 0;
> 
> _snakeForward: function(forward) {
> forward = (forward == 0 ? 0.00000000001 : forward);
> ...
> 
> Perhaps you can find a more elegant solution

I don't think you can get much better than that. Worked like a treat for me so thank you very much!!!

**Comment by boskoKlemenc** [comment]
> I found a work around by accounting for the "forward" variable in _snakeForward to not be 0;
> 
> _snakeForward: function(forward) {
> forward = (forward == 0 ? 0.00000000001 : forward);
> ...
> 
> Perhaps you can find a more elegant solution

It saved my day too. Thanks a lot !!!

**Comment by rfilmyer** [comment]
Is this worth including in an update/bugfix?

**Comment by Terrycoco** [comment]
I think so.  I haven’t had a prob since I added that fix

**Comment by rfilmyer** [comment]
I think I found the issue: 
`forward` comes from [this line](https://github.com/IvanSanchez/Leaflet.Polyline.SnakeAnim/blob/103991fe6876e8343a9f8737b2c12cf5c0799651/L.Polyline.SnakeAnim.js#L71). If `diff` ever equals 0, `forward` is 0, `distance` in `_snakeForward` is 0, and [0/0=`NaN`](https://github.com/IvanSanchez/Leaflet.Polyline.SnakeAnim/blob/103991fe6876e8343a9f8737b2c12cf5c0799651/L.Polyline.SnakeAnim.js#L114), so you get points with NaN x and y coordinates running around everywhere.

The code assumes that the difference between two `performance.now()` timestamps cannot ever be zero, because supposedly `performance.now()` should be so high-resolution that that could never happen.

Well, `performance.now()` [is not guaranteed to actually be high resolution](https://developer.mozilla.org/en-US/docs/Web/API/Performance/now). Firefox, in particular, has switched to giving it 1ms of resolution (was on the order of microseconds before) since version 60 (Spring 2018) for security reasons.

Clearly, [this call](https://github.com/IvanSanchez/Leaflet.Polyline.SnakeAnim/blob/103991fe6876e8343a9f8737b2c12cf5c0799651/L.Polyline.SnakeAnim.js#L43) and [this call](https://github.com/IvanSanchez/Leaflet.Polyline.SnakeAnim/blob/103991fe6876e8343a9f8737b2c12cf5c0799651/L.Polyline.SnakeAnim.js#L65) can happen in under a millisecond, meaning that the two timestamps are identical and their difference is 0.

@Terrycoco's fix works because it guarantees that `forward` can never be 0, but I brought the fix up to where the problem is in my pull request #37.
